### PR TITLE
Bluetooth: gatt: cancel all prepared writes on invalid write response

### PR DESCRIPTION
**Pull request by KKopyscinski**
On sending write request we should check write response.
If data prepaired to be written by peer or offset are not equal
to the data and offset we sent, we shall send Execute Write
Request with Request Flag set to 0x00 (Cancel All Prepared Writes).

This was affecting GATT/CL/GAW/BI-32-C and GATT/CL/GAW/BI-37-C

Signed-off-by: Krzysztof Kopyściński krzysztof.kopyscinski@codecoup.pl

### CODE REVIEW

File: subsys/bluetooth/host/gatt.c [hunk 1]
```diff
@@ -4122,11 +4143,20 @@ static void gatt_prepare_write_rsp(struct bt_conn *conn, uint8_t err,
 	}
 
 	len = length - sizeof(*rsp);
+
+	data_valid = memcmp(params->data, rsp->value, len) != 0;
```
Comment (joerchan) [inline subsys/bluetooth/host/gatt.c:4147]: Shouldn't use len before we have checked it against the params->length below. This should be moved down.

Comment (KKopyscinski) [reply subsys/bluetooth/host/gatt.c]: Of course, fixed

**Review by joerchan — APPROVED** [review]
Looks good. Please have a look at the compliance failure. There is a mismatch between commiter name/email and signed-off entry.

**Review by jhedberg — APPROVED** [review]
Looks fine, except for the Compliance check failure

**Comment by KKopyscinski** [comment]
@jhedberg Ok, fixed it. There was issue with polish signs in my surname

File: subsys/bluetooth/host/gatt.c [hunk 2]
```diff
@@ -4122,11 +4143,20 @@ static void gatt_prepare_write_rsp(struct bt_conn *conn, uint8_t err,
 	}
 
 	len = length - sizeof(*rsp);
+
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c:4146]: Oh, I missed this one: the style we've tried to follow in Bluetooth code is "no empty line between assignment to a variable and checking the variable's value". So please don't add this line.

File: subsys/bluetooth/host/gatt.c [hunk 3]
```diff
@@ -4122,11 +4143,20 @@ static void gatt_prepare_write_rsp(struct bt_conn *conn, uint8_t err,
...
 	if (len > params->length) {
 		params->func(conn, BT_ATT_ERR_INVALID_PDU, params);
 		return;
 	}
 
+	data_valid = memcmp(params->data, rsp->value, len) != 0;
+
```
Comment (jhedberg) [inline subsys/bluetooth/host/gatt.c:4153]: Likewise here you shouldn't need an empty line.

**Review by jhedberg — COMMENTED** [review]
Noticed a few minor coding style issues I failed to spot earlier. I'm not going to change my "Approved" review, but would be nice if you could address the issues.

**Comment by KKopyscinski** [comment]
@jhedberg Ok, I fixed coding style issues

**Comment by jhedberg** [comment]
@KKopyscinski looks good now, thanks!

**Comment by KKopyscinski** [comment]
Fixed mistake with `memcmp` error handling

**Comment by KKopyscinski** [comment]
@jhedberg @joerchan Added cancel writes if length is invalid, as bailing out still causes executing prepared writes. This was affecting GATT/CL/GAW/BI-37-C

**Comment by KKopyscinski** [comment]
All GATT/CL/GAW tests are passing

<details>
<summary>Test results</summary>

 1/23   GATT    GATT/CL/GAW/BV-01-C   PASS            37.697                  
 2/23   GATT    GATT/CL/GAW/BV-02-C   PASS            46.63                  
 3/23   GATT    GATT/CL/GAW/BV-03-C   PASS            37.439                  
 4/23   GATT    GATT/CL/GAW/BI-02-C   PASS            37.45                  
 5/23   GATT    GATT/CL/GAW/BI-03-C   PASS            37.675                  
 6/23   GATT    GATT/CL/GAW/BI-04-C   PASS            33.204                  
 7/23   GATT    GATT/CL/GAW/BI-05-C   PASS            35.131                  
 8/23   GATT    GATT/CL/GAW/BI-06-C   PASS            36.241                  
 9/23   GATT    GATT/CL/GAW/BV-05-C   PASS            37.324                  
10/23   GATT    GATT/CL/GAW/BI-07-C   PASS            36.858                  
11/23   GATT    GATT/CL/GAW/BI-08-C   PASS            37.059                  
12/23   GATT    GATT/CL/GAW/BI-09-C   PASS            36.984                  
13/23   GATT    GATT/CL/GAW/BI-11-C   PASS            35.363                  
14/23   GATT    GATT/CL/GAW/BI-12-C   PASS            36.203                  
15/23   GATT    GATT/CL/GAW/BI-13-C   PASS            36.122                  
16/23   GATT    GATT/CL/GAW/BV-06-C   PASS            38.025                  
17/23   GATT    GATT/CL/GAW/BV-08-C   PASS            37.044                  
18/23   GATT    GATT/CL/GAW/BV-09-C   PASS            37.103                  
19/23   GATT    GATT/CL/GAW/BI-32-C   PASS            37.078                  
20/23   GATT    GATT/CL/GAW/BI-33-C   PASS            36.978                  
21/23   GATT    GATT/CL/GAW/BI-34-C   PASS            37.242                  
22/23   GATT    GATT/CL/GAW/BV-10-C   PASS            30.135                  
23/23   GATT    GATT/CL/GAW/BI-37-C   PASS            30.091   

</details>

**Comment by KKopyscinski** [comment]
rebased onto main

**Pull request merged**